render/vulkan/shaders/common.vert

```diff
@@ -0,0 +1,25 @@
+#version 450
+
+// we use a mat4 since it uses the same size as mat3 due to
+// alignment. Easier to deal with (tighly-packed) mat4 though.
+layout(push_constant, row_major) uniform UBO {
```

Comment:
If you use std430, you can mirror such structures inside of C, since std430 is exactly the alignment C uses by default (unless you use compiler extensions, which don't even exist in regular POSIX).

Reply:
This block already is a push constant, which means it also already uses std430. Also, std430 mat3 is not tightly packed, it's equal to mat4.

Reply:
Right, but apart from matrices, it is close enough to C structs that you can map.